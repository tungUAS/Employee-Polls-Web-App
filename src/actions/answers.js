import {saveNewAnswer} from "../api";

export const RECEIVE_ANSWERS = "RECEIVE_ANSWERS";
export const ADD_ANSWER = "ADD_ANSWER";

export function receiveAnswers(answers) {
    return {
        type: RECEIVE_ANSWERS,
        answers,
    };
}

export function addNewAnswer(option, authedUser, questionId) {
    return {
        type: ADD_ANSWER,
        option,
        authedUser,
        questionId,
    };
};

export const handleAddNewAnswer = (option, authedUser, questionId) => {
    return async (dispatch) => {
        await saveNewAnswer(option, authedUser, questionId);
        dispatch(addNewAnswer(option, authedUser, questionId));
    };
};