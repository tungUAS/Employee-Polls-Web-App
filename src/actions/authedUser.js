export const SET_AUTHED_USER = 'SET_AUTHED_USER';

export function setAuthedUser(id,name) {
    return {
        type: SET_AUTHED_USER,
        id,
        name
    };
}