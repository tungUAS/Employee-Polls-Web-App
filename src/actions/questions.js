import { saveNewQuestion, saveAnsweredByQuestion } from "../api";

export const RECEIVE_QUESTIONS = "RECEIVE_QUESTIONS";
export const ADD_ANSWERED_BY_QUESTION = "ADD_ANSWERED_BY_QUESTION";
export const ADD_NEW_QUESTION = "ADD_NEW_QUESTION";

export const receiveQuestions = (questions) => {
  return {
    type: RECEIVE_QUESTIONS,
    questions,
  };
};

export const addAnsweredByQuestion = (authedUserId, questionId) => {
  return {
    type: ADD_ANSWERED_BY_QUESTION,
    authedUserId,
    questionId,
  };
};

export const addNewQuestion = (authedUserId, optionOne, optionTwo) => {
  return {
    type: ADD_NEW_QUESTION,
    authedUserId,
    optionOne,
    optionTwo,
  };
}

export const handleAddAnsweredByQuestion = (authedUserId, questionId) => {
  return async (dispatch) => {
    dispatch(addAnsweredByQuestion(authedUserId, questionId));
    await saveAnsweredByQuestion(authedUserId, questionId);
  };
}

export const handleAddNewQuestion = (authedUserId, optionOne, optionTwo) => {
  return async (dispatch) => {
    dispatch(addNewQuestion(authedUserId, optionOne, optionTwo));
    await saveNewQuestion(authedUserId, optionOne, optionTwo);
  };
}