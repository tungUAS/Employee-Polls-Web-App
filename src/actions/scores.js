export const RECEIVE_SCORES = "RECEIVE_SCORES";
export const UPDATE_ANSWERED_SCORES = "UPDATE_ANSWERED_SCORES";
export const UPDATE_CREATED_SCORES = "UPDATE_CREATED_SCORES";

export const receiveScores = (scores) => {
    return {
        type: RECEIVE_SCORES,
        scores,
    };
};

export const updateAnsweredScores = (authedUserId) => {
    return {
        type: UPDATE_ANSWERED_SCORES,
        authedUserId,
    };
}

export const updateCreatedScores = (authedUserId) => {
    return {
        type: UPDATE_CREATED_SCORES,
        authedUserId,
    };
}