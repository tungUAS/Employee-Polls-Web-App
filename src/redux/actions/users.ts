import { UsersType } from "../../models/user.type";

export const RECEIVE_USERS = "RECEIVE_USERS";

export const receiveUsers = ({users}:{users: UsersType}) => {
  return {
    type: RECEIVE_USERS,
    users,
  };
};