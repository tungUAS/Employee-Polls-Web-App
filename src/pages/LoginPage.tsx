import React, { useState } from "react";
import { connect } from "react-redux";
import { setAuthedUser } from "../redux/actions/authedUser";
import { useNavigate } from "react-router-dom";
import { UsersType } from "../models/user.type";
import { Dispatch } from "redux";
import "../styles/Login.css";

const LoginPage = ({
  users,
  dispatch,
}: {
  users: UsersType;
  dispatch: Dispatch;
}) => {
  const [selectedName, setSelectedName] = useState("Sarah");
  const navigate = useNavigate();

  const pathToBeRedirected = localStorage.getItem("redirect");

  const handleChange = (e: React.ChangeEvent<HTMLSelectElement>) => {
    setSelectedName(e.target.value);
  };

  const handleLogin = () => {
    if(pathToBeRedirected){
      navigate(pathToBeRedirected);
    };

    const user = users.find((user) => user.name === selectedName);
    if (!user) return;
    dispatch(
      setAuthedUser({
        id: user.id,
        name: user.name,
      })
    );
    navigate(pathToBeRedirected ? pathToBeRedirected : "/questions");
  };

  return (
    <div className="login-container">
      <h1>Log In</h1>
      <img
        src="https://gravatar.com/avatar/17461862a35eb112646cd5f27097ca85?s=400&d=robohash&r=x"
        alt="Login Background"
      />
      <div>
        <label htmlFor="name">Select your name: </label>
        <select id="name" value={selectedName} onChange={handleChange}>
          {users.map((user) => (
            <option key={user.id} value={user.name} data-testid="select-option">
              {user.name}
            </option>
          ))}
        </select>
      </div>
      <br />
      <br />
      <button onClick={handleLogin} className="login-button">
        Let's Go
      </button>
    </div>
  );
};

const mapStateToProps = ({ users }: { users: UsersType }) => ({
  users,
});

export default connect(mapStateToProps)(LoginPage);
