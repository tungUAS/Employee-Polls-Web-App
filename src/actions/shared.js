import { getInitialData } from "../api";
import { receiveQuestions } from "./questions";
import { receiveUsers } from "./users";
import { receiveScores } from "./scores";
import { receiveAnswers } from "./answers";

export function handleInitialData() {
  return (dispatch) => {
    return getInitialData().then(({ questions, users, scores, answers }) => {
      dispatch(receiveQuestions(questions));
      dispatch(receiveUsers(users));
      dispatch(receiveScores(scores));
      dispatch(receiveAnswers(answers));
    });
  };
}
