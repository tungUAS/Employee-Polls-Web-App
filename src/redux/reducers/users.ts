import { UsersType } from "../../models/user.type";
import { RECEIVE_USERS } from "../actions/users";

const users = (state:UsersType = [], action: {
  type: string;
  users: UsersType;
}) => {
  switch (action.type) {
    case RECEIVE_USERS:
      return [
        ...state,
        ...action.users,
      ];
    default:
      return state;
  }
};

export default users;