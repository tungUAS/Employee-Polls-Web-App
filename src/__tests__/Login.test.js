import React from "react";
import { fireEvent, render } from "@testing-library/react";
import LoginPage from "../pages/LoginPage";
import { Provider } from "react-redux";
import { Store, legacy_createStore as createStore } from "redux";
import reducer from "../redux/reducers";
import { MemoryRouter } from "react-router-dom";
import { receiveUsers } from "../redux/actions/users";
import { UsersType } from "../models/user.type";

const mockNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockNavigate,
}));

describe("Login Page Test", () => {
  const mockUsers = [
    { id: 1, name: "Sarah", avatar_url: "avatar1" },
    { id: 2, name: "Tyler", avatar_url: "avatar2" },
  ];

  it("renders login form with select input and button", () => {
    const store = createStore(reducer);
    store.dispatch(receiveUsers({users: mockUsers}));

    const { getByText, getByLabelText, getByRole, asFragment } = render(
      <MemoryRouter>
        <Provider store={store}>
          <LoginPage users={mockUsers} />
        </Provider>
      </MemoryRouter>
    );

    const loginHeader = getByText("Log In");
    expect(loginHeader).toBeInTheDocument();

    const selectInput = getByLabelText("Select your name:");
    expect(selectInput).toBeInTheDocument();

    const loginButton = getByRole("button", { name: "Let's Go" });
    expect(loginButton).toBeInTheDocument();

    expect(asFragment()).toMatchSnapshot();
  });

  it("dispatches setAuthedUser action and navigates to questions page on login", () => {
    const store = createStore(reducer);
    store.dispatch(receiveUsers({users: mockUsers}));

    const { getByLabelText, getAllByTestId, getByRole } = render(
      <MemoryRouter>
        <Provider store={store}>
          <LoginPage users={mockUsers} />
        </Provider>
      </MemoryRouter>
    );

    const selectInput = getByLabelText("Select your name:");
    fireEvent.change(selectInput, { target: { value: "Sarah" } });
    const options = getAllByTestId("select-option");
    expect(options[0].selected).toBe(true);

    const loginButton = getByRole("button", { name: "Let's Go" });
    fireEvent.click(loginButton);
    setTimeout(() => {
      expect(store.getState().authedUser).toEqual({
        id: 1,
        name: "Sarah",
      });

      expect(mockNavigate).toBe("/questions");
    }, 1000);
  });

  it("should not dispatch setAuthedUser action and navigates to questions page on login", () => {
    const store = createStore(reducer);

    render(
        <MemoryRouter>
        <Provider store={store}>
          <LoginPage users={mockUsers} />
        </Provider>
      </MemoryRouter>
    );

    setTimeout(() => {
      expect(store.getState().authedUser).toEqual(null);

      expect(mockNavigate).notToHaveBeenCalled();
    }, 1000);
  });
});
