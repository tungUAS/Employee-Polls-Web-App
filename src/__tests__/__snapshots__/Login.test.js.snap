// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Page Test renders login form with select input and button 1`] = `
<DocumentFragment>
  <div
    class="login-container"
  >
    <h1>
      Log In
    </h1>
    <img
      alt="Login Background"
      src="https://gravatar.com/avatar/17461862a35eb112646cd5f27097ca85?s=400&d=robohash&r=x"
    />
    <div>
      <label
        for="name"
      >
        Select your name: 
      </label>
      <select
        id="name"
      >
        <option
          data-testid="select-option"
          value="Sarah"
        >
          Sarah
        </option>
        <option
          data-testid="select-option"
          value="Tyler"
        >
          Tyler
        </option>
      </select>
    </div>
    <br />
    <br />
    <button
      class="login-button"
    >
      Let's Go
    </button>
  </div>
</DocumentFragment>
`;
